### PR TITLE
JAMES-2368 Allow James to register listener

### PR DESCRIPTION
**Pull request by aduprat**

### CODE REVIEW

File: server/container/guice/mailbox/src/main/java/org/apache/james/modules/mailbox/NoopMailboxListener.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import org.apache.james.mailbox.Event;
+import org.apache.james.mailbox.MailboxListener;
+import org.apache.james.mailbox.MailboxListener.ExecutionMode;
+import org.apache.james.mailbox.MailboxListener.ListenerType;
+
+public class NoopMailboxListener implements MailboxListener {
+
+    @Override
+    public ListenerType getType() {
+        return null;
```
Comment (rouazana) [inline server/container/guice/mailbox/src/main/java/org/apache/james/modules/mailbox/NoopMailboxListener.java]: throw NotImplemented?

File: server/container/guice/mailbox/src/main/java/org/apache/james/modules/mailbox/NoopMailboxListener.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    }
+
+    @Override
+    public ExecutionMode getExecutionMode() {
+        return null;
```
Comment (rouazana) [inline server/container/guice/mailbox/src/main/java/org/apache/james/modules/mailbox/NoopMailboxListener.java]: idem?

File: dockerfiles/run/guice/cassandra/destination/conf/listeners.xml [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  <listener>
+    <class>org.apache.james.jmap.event.PropagateLookupRightListener</class>
+  </listener>
+  <listener>
+    <class>org.apache.james.mailbox.spamassassin.SpamAssassinListener</class>
+  </listener>
+  <listener>
+    <class>org.apache.james.mailbox.cassandra.MailboxOperationLoggingListener</class>
+  </listener>
+</listeners>
```
Comment (chibenwa) [inline dockerfiles/run/guice/cassandra/destination/conf/listeners.xml]: Are you missing this for Guice-JPA files?

**Comment by chibenwa** [comment]
In Cassandra JMAP

```
[3192b95a206d86669caf014578605124f4ce9fb2] Results :
[3192b95a206d86669caf014578605124f4ce9fb2] 
[3192b95a206d86669caf014578605124f4ce9fb2] Failed tests: 
[3192b95a206d86669caf014578605124f4ce9fb2]   CassandraMailboxSharingTest 
[3192b95a206d86669caf014578605124f4ce9fb2] Expecting:
[3192b95a206d86669caf014578605124f4ce9fb2]   <[null]>
[3192b95a206d86669caf014578605124f4ce9fb2] to contain only:
[3192b95a206d86669caf014578605124f4ce9fb2]   <["de4310f0-4c97-11e8-bb3c-cf9a1f960175"]>
[3192b95a206d86669caf014578605124f4ce9fb2] elements not found:
[3192b95a206d86669caf014578605124f4ce9fb2]   <["de4310f0-4c97-11e8-bb3c-cf9a1f960175"]>
[3192b95a206d86669caf014578605124f4ce9fb2] and elements not expected:
[3192b95a206d86669caf014578605124f4ce9fb2]   <[null]>
[3192b95a206d86669caf014578605124f4ce9fb2] 
[3192b95a206d86669caf014578605124f4ce9fb2]   CassandraMailboxSharingTest 
[3192b95a206d86669caf014578605124f4ce9fb2] Expecting:
[3192b95a206d86669caf014578605124f4ce9fb2]  <["INBOX", "Drafts", "Outbox", "Sent", "Trash", "Spam", "shared", "shared"]>
[3192b95a206d86669caf014578605124f4ce9fb2] to contain:
[3192b95a206d86669caf014578605124f4ce9fb2]  <["mailbox1"]>
[3192b95a206d86669caf014578605124f4ce9fb2] but could not find:
[3192b95a206d86669caf014578605124f4ce9fb2]  <["mailbox1"]>
[3192b95a206d86669caf014578605124f4ce9fb2] 
[3192b95a206d86669caf014578605124f4ce9fb2]   CassandraMailboxSharingTest 
[3192b95a206d86669caf014578605124f4ce9fb2] Expecting:
[3192b95a206d86669caf014578605124f4ce9fb2]   <[null]>
[3192b95a206d86669caf014578605124f4ce9fb2] to contain only:
[3192b95a206d86669caf014578605124f4ce9fb2]   <["de4310f0-4c97-11e8-bb3c-cf9a1f960175"]>
[3192b95a206d86669caf014578605124f4ce9fb2] elements not found:
[3192b95a206d86669caf014578605124f4ce9fb2]   <["de4310f0-4c97-11e8-bb3c-cf9a1f960175"]>
[3192b95a206d86669caf014578605124f4ce9fb2] and elements not expected:
[3192b95a206d86669caf014578605124f4ce9fb2]   <[null]>
[3192b95a206d86669caf014578605124f4ce9fb2] 
[3192b95a206d86669caf014578605124f4ce9fb2]   CassandraMailboxSharingTest 
[3192b95a206d86669caf014578605124f4ce9fb2] Expecting:
[3192b95a206d86669caf014578605124f4ce9fb2]  <["INBOX", "Drafts", "Outbox", "Sent", "Trash", "Spam", "shared", "shared"]>
[3192b95a206d86669caf014578605124f4ce9fb2] to contain:
[3192b95a206d86669caf014578605124f4ce9fb2]  <["mailbox1"]>
[3192b95a206d86669caf014578605124f4ce9fb2] but could not find:
[3192b95a206d86669caf014578605124f4ce9fb2]  <["mailbox1"]>
[3192b95a206d86669caf014578605124f4ce9fb2] 
[3192b95a206d86669caf014578605124f4ce9fb2] Tests in error: 
[3192b95a206d86669caf014578605124f4ce9fb2]   SpamAssassinContract.deletingSpamMessageShouldNotImpactSpamAssassinState:538 » ConditionTimeout
[3192b95a206d86669caf014578605124f4ce9fb2]   SpamAssassinContract.expungingSpamMessageShouldNotImpactSpamAssassinState:475 » ConditionTimeout
[3192b95a206d86669caf014578605124f4ce9fb2]   SpamAssassinContract.imapCopiesToSpamMailboxShouldBeConsideredAsSpam:181 » ConditionTimeout
[3192b95a206d86669caf014578605124f4ce9fb2]   SpamAssassinContract.imapMovesToSpamMailboxShouldBeConsideredAsSpam:228 » ConditionTimeout
[3192b95a206d86669caf014578605124f4ce9fb2]   SpamAssassinContract.movingAMailToTrashShouldNotImpactSpamassassinLearning:354 » ConditionTimeout
[3192b95a206d86669caf014578605124f4ce9fb2]   SpamAssassinContract.spamShouldBeDeliveredInSpamMailboxOrInboxWhenMultipleRecipientsConfigurations:608 » ConditionTimeout
[3192b95a206d86669caf014578605124f4ce9fb2]   SpamAssassinContract.spamShouldBeDeliveredInSpamMailboxWhenSameMessageHasAlreadyBeenMovedToSpam:134 » ConditionTimeout
```

I'd bet "missing configuration files".

**Comment by aduprat** [comment]
:+1:

File: src/site/xdoc/server/config-listeners.xml [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        <title>Apache James Server 3 - Listeners Configuration</title>
+    </properties>
+
+    <body>
+
+        <section name="Mailbox listeners Configuration">
+
+            <p>
+              Note: This feature requires Guice wiring. This is not implemented using Spring wiring.<br/>
+              Additionally wiring not providing a mailbox support, like <code>jpa-smtp</code>, do not have support for
```
Comment (rouazana) [inline src/site/xdoc/server/config-listeners.xml]: Additional?

File: server/container/guice/mailbox/src/main/java/org/apache/james/modules/mailbox/GlobalMailboxListeners.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+    @VisibleForTesting void configureListener(HierarchicalConfiguration configuration) {
+        String listenerClass = configuration.getString("class");
+        Preconditions.checkState(!Strings.isNullOrEmpty(listenerClass), "class name is mandatory");
+        try {
+            LOGGER.info("Loading mailbox listener {}", listenerClass);
+            registry.addGlobalListener(injector.getInstance(loadMailboxListener(listenerClass)));
+        } catch (ClassNotFoundException | InstantiationException | IllegalAccessException | MailboxException e) {
+            LOGGER.error("Error while loading global listener {}", listenerClass, e);
+            Throwables.propagate(e);
```
Comment (mbaechler) [inline server/container/guice/mailbox/src/main/java/org/apache/james/modules/mailbox/GlobalMailboxListeners.java]: please don't use that anymore, the right solution is now `throw new RuntimeException(e);`

File: server/container/guice/mailbox/src/main/java/org/apache/james/modules/mailbox/GlobalMailboxListeners.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+    @Override
+    public void configure(HierarchicalConfiguration configuration) throws ConfigurationException {
+        LOGGER.info("Loading mailbox listeners");
+        List<HierarchicalConfiguration> listenersConfiguration = configuration.configurationsAt("listener");
+        listenersConfiguration.stream()
+            .forEach(listenerConfiguration -> configureListener(listenerConfiguration));
+    }
+
+    @VisibleForTesting void configureListener(HierarchicalConfiguration configuration) {
```
Comment (mbaechler) [inline server/container/guice/mailbox/src/main/java/org/apache/james/modules/mailbox/GlobalMailboxListeners.java]: for new code I'd like we define configuration as POJO and have a reader that goes from HierarchicalConfiguration to that POJO

File: server/container/guice/mailbox/src/main/java/org/apache/james/modules/mailbox/GlobalMailboxListeners.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+public class GlobalMailboxListeners implements Configurable {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(GlobalMailboxListeners.class);
+
+    private final Injector injector;
+    private final MailboxListenerRegistry registry;
+
+    @Inject
+    public GlobalMailboxListeners(Injector injector, MailboxListenerRegistry registry) {
```
[the remark this answers is not in the document]
Comment (aduprat) [reply server/container/guice/mailbox/src/main/java/org/apache/james/modules/mailbox/GlobalMailboxListeners.java]: There's no rule about listener constructor.
Some of them needs things to be injected:
```
    @Inject
    public SpamAssassinListener(SpamAssassin spamAssassin, MailboxSessionMapperFactory mapperFactory) {
```
then, it's hard to instantiate them.

That's why I take the choice to use the _injector_.

This class is only responsible on extra global listeners registration, not on they should be instantiated in a guice module.

[the remark this answers is not in the document]
Comment (chibenwa) [reply server/container/guice/mailbox/src/main/java/org/apache/james/modules/mailbox/GlobalMailboxListeners.java]: I guess @mbaechler  asks for something like `server/mailet/mailetcontainer-api/src/main/java/org/apache/james/mailetcontainer/api/MailetLoader.java`... There is no conflict instantiating them threw Guice...

**Comment by aduprat** [comment]
```
Tests in error: 
  SpamAssassinContract.deletingSpamMessageShouldNotImpactSpamAssassinState:538 » ConditionTimeout
  SpamAssassinContract.expungingSpamMessageShouldNotImpactSpamAssassinState:475 » ConditionTimeout
  SpamAssassinContract.imapCopiesToSpamMailboxShouldBeConsideredAsSpam:181 » ConditionTimeout
  SpamAssassinContract.imapMovesToSpamMailboxShouldBeConsideredAsSpam:228 » ConditionTimeout
  SpamAssassinContract.movingAMailToTrashShouldNotImpactSpamassassinLearning:354 » ConditionTimeout
  SpamAssassinContract.spamShouldBeDeliveredInSpamMailboxOrInboxWhenMultipleRecipientsConfigurations:608 » ConditionTimeout
  SpamAssassinContract.spamShouldBeDeliveredInSpamMailboxWhenSameMessageHasAlreadyBeenMovedToSpam:134 » ConditionTimeout
```

**Comment by chibenwa** [comment]
```
[40b938b84d194704ffeba401ec4f8a47672a862d] Results :
[40b938b84d194704ffeba401ec4f8a47672a862d] 
[40b938b84d194704ffeba401ec4f8a47672a862d] Failed tests: 
[40b938b84d194704ffeba401ec4f8a47672a862d]   CassandraMailboxSharingTest 
[40b938b84d194704ffeba401ec4f8a47672a862d] Expecting:
[40b938b84d194704ffeba401ec4f8a47672a862d]   <[null]>
[40b938b84d194704ffeba401ec4f8a47672a862d] to contain only:
[40b938b84d194704ffeba401ec4f8a47672a862d]   <["a4d77520-4e1a-11e8-8e5a-a163941f16bb"]>
[40b938b84d194704ffeba401ec4f8a47672a862d] elements not found:
[40b938b84d194704ffeba401ec4f8a47672a862d]   <["a4d77520-4e1a-11e8-8e5a-a163941f16bb"]>
[40b938b84d194704ffeba401ec4f8a47672a862d] and elements not expected:
[40b938b84d194704ffeba401ec4f8a47672a862d]   <[null]>
[40b938b84d194704ffeba401ec4f8a47672a862d] 
[40b938b84d194704ffeba401ec4f8a47672a862d]   CassandraMailboxSharingTest 
[40b938b84d194704ffeba401ec4f8a47672a862d] Expecting:
[40b938b84d194704ffeba401ec4f8a47672a862d]  <["INBOX", "Drafts", "Outbox", "Sent", "Trash", "Spam", "shared", "shared"]>
[40b938b84d194704ffeba401ec4f8a47672a862d] to contain:
[40b938b84d194704ffeba401ec4f8a47672a862d]  <["mailbox1"]>
[40b938b84d194704ffeba401ec4f8a47672a862d] but could not find:
[40b938b84d194704ffeba401ec4f8a47672a862d]  <["mailbox1"]>
[40b938b84d194704ffeba401ec4f8a47672a862d] 
[40b938b84d194704ffeba401ec4f8a47672a862d]   CassandraMailboxSharingTest 
[40b938b84d194704ffeba401ec4f8a47672a862d] Expecting:
[40b938b84d194704ffeba401ec4f8a47672a862d]   <[null]>
[40b938b84d194704ffeba401ec4f8a47672a862d] to contain only:
[40b938b84d194704ffeba401ec4f8a47672a862d]   <["a4d77520-4e1a-11e8-8e5a-a163941f16bb"]>
[40b938b84d194704ffeba401ec4f8a47672a862d] elements not found:
[40b938b84d194704ffeba401ec4f8a47672a862d]   <["a4d77520-4e1a-11e8-8e5a-a163941f16bb"]>
[40b938b84d194704ffeba401ec4f8a47672a862d] and elements not expected:
[40b938b84d194704ffeba401ec4f8a47672a862d]   <[null]>
[40b938b84d194704ffeba401ec4f8a47672a862d] 
[40b938b84d194704ffeba401ec4f8a47672a862d]   CassandraMailboxSharingTest 
[40b938b84d194704ffeba401ec4f8a47672a862d] Expecting:
[40b938b84d194704ffeba401ec4f8a47672a862d]  <["INBOX", "Drafts", "Outbox", "Sent", "Trash", "Spam", "shared", "shared"]>
[40b938b84d194704ffeba401ec4f8a47672a862d] to contain:
[40b938b84d194704ffeba401ec4f8a47672a862d]  <["mailbox1"]>
[40b938b84d194704ffeba401ec4f8a47672a862d] but could not find:
[40b938b84d194704ffeba401ec4f8a47672a862d]  <["mailbox1"]>
[40b938b84d194704ffeba401ec4f8a47672a862d] 
[40b938b84d194704ffeba401ec4f8a47672a862d] Tests in error: 
[40b938b84d194704ffeba401ec4f8a47672a862d]   SpamAssassinContract.deletingSpamMessageShouldNotImpactSpamAssassinState:538 » ConditionTimeout
[40b938b84d194704ffeba401ec4f8a47672a862d]   SpamAssassinContract.expungingSpamMessageShouldNotImpactSpamAssassinState:475 » ConditionTimeout
[40b938b84d194704ffeba401ec4f8a47672a862d]   SpamAssassinContract.imapCopiesToSpamMailboxShouldBeConsideredAsSpam:181 » ConditionTimeout
[40b938b84d194704ffeba401ec4f8a47672a862d]   SpamAssassinContract.imapMovesToSpamMailboxShouldBeConsideredAsSpam:228 » ConditionTimeout
[40b938b84d194704ffeba401ec4f8a47672a862d]   SpamAssassinContract.movingAMailToTrashShouldNotImpactSpamassassinLearning:354 » ConditionTimeout
[40b938b84d194704ffeba401ec4f8a47672a862d]   SpamAssassinContract.spamShouldBeDeliveredInSpamMailboxOrInboxWhenMultipleRecipientsConfigurations:608 » ConditionTimeout
[40b938b84d194704ffeba401ec4f8a47672a862d]   SpamAssassinContract.spamShouldBeDeliveredInSpamMailboxWhenSameMessageHasAlreadyBeenMovedToSpam:134 » ConditionTimeout
[40b938b84d194704ffeba401ec4f8a47672a862d] 
```

Still... I'll have a look...

**Comment by chibenwa** [comment]
```
[0f797568bf9811b979d6a5e978a0315e52683317] Tests in error: 
[0f797568bf9811b979d6a5e978a0315e52683317]   SpamAssassinContract.imapCopiesToSpamMailboxShouldBeConsideredAsSpam:181 » ConditionTimeout
```

test this please

File: server/container/guice/mailbox/src/main/java/org/apache/james/modules/mailbox/MailboxListenersLoaderImpl.java [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(MailboxListenersLoaderImpl.class);
+
+    private final Injector injector;
+    private final MailboxListenerRegistry registry;
+    private final ExtendedClassLoader classLoader;
+    private final Set<MailboxListener> guiceDefinedListeners;
+
+    @Inject
+    public MailboxListenersLoaderImpl(Injector injector, MailboxListenerRegistry registry,
```
Comment (mbaechler) [inline server/container/guice/mailbox/src/main/java/org/apache/james/modules/mailbox/MailboxListenersLoaderImpl.java]: Sorry if I was unclear about Injector use.
Here, you inject  `Injector`  for a single call `MailboxListener listener = injector.getInstance(clazz);`
In my opinion, it deserves a class `MailboxListenerFactory` with a single method `MailboxListener createInstance(clazz)` that would rely on the Injector.

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
```
[4df336d9eed8302a00b26864723abd1c62bbaf03] Tests run: 66, Failures: 1, Errors: 0, Skipped: 7, Time elapsed: 50.102 sec <<< FAILURE! - in org.apache.james.queue.activemq.ActiveMQMailQueueBlobTest
[4df336d9eed8302a00b26864723abd1c62bbaf03] concurrentEnqueueShouldNotAlterBrowsing()  Time elapsed: 0.031 sec  <<< FAILURE!
[4df336d9eed8302a00b26864723abd1c62bbaf03] java.lang.AssertionError: 
[4df336d9eed8302a00b26864723abd1c62bbaf03] 
[4df336d9eed8302a00b26864723abd1c62bbaf03] Actual and expected should have same size but actual size was:
[4df336d9eed8302a00b26864723abd1c62bbaf03]   <3>
[4df336d9eed8302a00b26864723abd1c62bbaf03] while expected size was:
[4df336d9eed8302a00b26864723abd1c62bbaf03]   <2>
[4df336d9eed8302a00b26864723abd1c62bbaf03] Actual was:
[4df336d9eed8302a00b26864723abd1c62bbaf03]   <["name1", "name2", "name3"]>
[4df336d9eed8302a00b26864723abd1c62bbaf03] Expected was:
[4df336d9eed8302a00b26864723abd1c62bbaf03]   <["name1", "name2"]>
[4df336d9eed8302a00b26864723abd1c62bbaf03] 
[4df336d9eed8302a00b26864723abd1c62bbaf03] 	at org.apache.james.queue.api.ManageableMailQueueContract.concurrentEnqueueShouldNotAlterBrowsing(ManageableMailQueueContract.java:245)
[4df336d9eed8302a00b26864723abd1c62bbaf03] 
[4df336d9eed8302a00b26864723abd1c62bbaf03] 
[4df336d9eed8302a00b26864723abd1c62bbaf03] Results :
[4df336d9eed8302a00b26864723abd1c62bbaf03] 
[4df336d9eed8302a00b26864723abd1c62bbaf03] Failed tests: 
[4df336d9eed8302a00b26864723abd1c62bbaf03]   ManageableMailQueueContract.concurrentEnqueueShouldNotAlterBrowsing:245 
[4df336d9eed8302a00b26864723abd1c62bbaf03] Actual and expected should have same size but actual size was:
[4df336d9eed8302a00b26864723abd1c62bbaf03]   <3>
[4df336d9eed8302a00b26864723abd1c62bbaf03] while expected size was:
[4df336d9eed8302a00b26864723abd1c62bbaf03]   <2>
[4df336d9eed8302a00b26864723abd1c62bbaf03] Actual was:
[4df336d9eed8302a00b26864723abd1c62bbaf03]   <["name1", "name2", "name3"]>
[4df336d9eed8302a00b26864723abd1c62bbaf03] Expected was:
[4df336d9eed8302a00b26864723abd1c62bbaf03]   <["name1", "name2"]>
[4df336d9eed8302a00b26864723abd1c62bbaf03] 
[4df336d9eed8302a00b26864723abd1c62bbaf03] 
[4df336d9eed8302a00b26864723abd1c62bbaf03] Tests run: 132, Failures: 1, Errors: 0, Skipped: 14
```

test this please

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**